python-checks/src/main/java/org/sonar/python/checks/Expressions.java

```diff
@@ -103,4 +108,138 @@ public static Expression removeParentheses(@Nullable Expression expression) {
+
+    private static final EscapeSequence IGNORE = new EscapeSequence(1, "\\");
+
+    private static final char[] UNESCAPED_CHAR = new char['v' + 1];
+    static {
+      UNESCAPED_CHAR['\\'] = '\\';
+      UNESCAPED_CHAR['\''] = '\'';
+      UNESCAPED_CHAR['\"'] = '\"';
+      UNESCAPED_CHAR['a'] = '\u0007';
+      UNESCAPED_CHAR['b'] = '\b';
+      UNESCAPED_CHAR['f'] = '\f';
+      UNESCAPED_CHAR['n'] = '\n';
+      UNESCAPED_CHAR['r'] = '\r';
+      UNESCAPED_CHAR['t'] = '\t';
+      UNESCAPED_CHAR['v'] = '\u000b';
+    }
+
+    private final int escapedLength;
+    private final String unescapedValue;
+
+    private EscapeSequence(int escapedLength, String unescapedValue) {
+      this.escapedLength = escapedLength;
+      this.unescapedValue = unescapedValue;
+    }
+
+    private static EscapeSequence extract(String value, int i, boolean support16Bits) {
+      if (i == value.length() - 1) {
+        return IGNORE;
+      }
+      char nextChar = value.charAt(i + 1);
+      char unescaped = nextChar < UNESCAPED_CHAR.length ? UNESCAPED_CHAR[nextChar] : '\0';
+      if (unescaped != '\0') {
+        return new EscapeSequence(2, String.valueOf(unescaped));
+      } else if (nextChar == '\n') {
+        // ignored line break (linux end of line)
+        return new EscapeSequence(2, "");
+      } else if (nextChar == '\r') {
+        // ignored line break (windows and mac end of line)
+        return new EscapeSequence(i + 2 < value.length() && value.charAt(i + 2) == '\n' ? 3 : 2, "");
+      } else if (nextChar == 'x') {
+        return extractHexadecimal(value, i, 2);
+      } else if (nextChar == 'u' && support16Bits) {
+        return extractHexadecimal(value, i, 4);
+      } else if (nextChar == 'U' && support16Bits) {
+        return extractHexadecimal(value, i, 8);
+      } else if (nextChar == 'N') {
+        // escape sequence by unicode name is not supported, require java 9 to benefit from Character.codePointOf
+        return IGNORE;
+      } else {
+        return extractOctal(value, i);
+      }
+    }
+
+    private static EscapeSequence extractHexadecimal(String value, int i, int length) {
+      if (i + 1 + length < value.length()) {
+        try {
+          int hexValue = Integer.parseInt(value.substring(i + 2, i + 2 + length), HEXADECIMAL_RADIX);
+          return new EscapeSequence(2 + length, String.valueOf((char) hexValue));
+        } catch (NumberFormatException ex) {
+          return IGNORE;
+        }
+      } else {
+        return IGNORE;
+      }
+    }
+
+    private static EscapeSequence extractOctal(String value, int i) {
+      // octal
+      int octal = 0;
+      int octalStart = (value.charAt(i + 1) == 'o') ? (i + 2) : (i + 1);
+      int len = 0;
+      int j = octalStart;
+      while (len < 3 && j < value.length() && value.charAt(j) >= '0' && value.charAt(j) <= '7') {
+        octal = octal * 8 + (value.charAt(j) - '0');
+        j++;
+        len++;
+      }
```

Comment:
Why not using `Integer.parseInt` the same way you did it for `extractHexadecimal`? Or maybe even sharing code for the parsing of octal and hexadecimal...

Reply:
- for hexadecimal, we know the length in advance (x:2 u:4 U:8)
- for octal it's a mess, 'o' prefix is not mandatory, the length can be from 1 to 3, we stop at 3 or when the next character does not match 0-7